### PR TITLE
Fix conversion issues in null coalescing operator

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/60046 (the issue was that `ApplyConversionIfNotIdentity` didn't consider scenario with identity conversion on top of another conversion)
Fixes https://github.com/dotnet/roslyn/issues/60059 and fixes https://github.com/dotnet/roslyn/issues/58176 (the issue was that initial binding created a conversion from nullable type to result type, instead of from underlying type to result type)

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:3846]: :memo: When the LHS of null-coalescing operator is a nullable value type, in the branch where LHS is not null, the lowering does `.GetValueOrDefault` before converting to the result type. See `GetConvertedLeftForNullCoalescingOperator`.
In the case of `dynamic` as result type, this mean we'll do `(object)left.GetValueOrDefault`, so the conversion here needs to be `leftType0 -> object`.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review. Thanks

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:3846]: >optLeftType0 [](http://example.com/codeflow?start=84&length=12)

This change looks suspicious. The type doesn't match the conversion that we are classifying below.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:240]: >ConsiderEverything [](http://example.com/codeflow?start=91&length=18)

What analysis was performed to make this change? If this is just a mechanical change, we should undo it because the point of using ```ConsiderEverything2``` was to come back and think about what kind should actually be used.

File: src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:4937]: >NullableLongAndInt [](http://example.com/codeflow?start=37&length=18)

Consider also adding a test for 
```
int? l = 42;
long i = 43;
System.Console.Write(l ?? i);
```

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:240]: >if (!TypeSymbol.Equals(rewrittenLeftType, rewrittenResultType, TypeCompareKind.ConsiderEverything) && rewrittenLeftType.IsNullableType()) [](http://example.com/codeflow?start=12&length=137)

Is getting to this ```if``` causing the unwrapping? If so, perhaps we could add a check whether placeholder's type is Nullable and prevent unwrapping in that case.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: You're right. I missed that we're still using `leftOperand` in `ClassifyConversionFromExpression`, so they don't match :-/

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs]: That is the point of `ConsiderEverything2`. I think `ConsiderEverything` is correct here but given this is not central to this PR, I'm reverting.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs]: > That is the point of ConsiderEverything2. I think ConsiderEverything is correct here 

This isn't that obvious to me. It looks like we are concern about identity convertibility between the types here. In cases like that ignoring everything (nullable annotations, dynamic vs. object, custom modifiers, ...) feels more appropriate. Changes like this, usually, should be accompanied by tests proving that the used option is appropriate and necessary.

**Comment by jcouv** [comment]
Looking at other affected tests...

File: src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:4966]: > ""long long?.Value.get"" [](http://example.com/codeflow?start=22&length=25)

This change, though correct behavior-wise, indicates that we started generating less optimal code. The code though exactly matches what we produce during binding. We indeed create this conversion. I think I have an idea how to address it. Will respond to this thread shortly

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:4966]: Here is the problematic code in ```BindNullCoalescingOperator```:
```
            if (isLeftNullable)
            {
                var rightConversion = Conversions.ClassifyImplicitConversionFromExpression(rightOperand, optLeftType0, ref useSiteInfo);
                if (rightConversion.Exists)
                {
                    var leftPlaceholder = new BoundValuePlaceholder(leftOperand.Syntax, optLeftType).MakeCompilerGenerated();
                    var leftConversion = CreateConversion(node, leftPlaceholder,
                                                          Conversions.ClassifyConversionFromExpression(leftOperand, optLeftType0, ref useSiteInfo),
                                                          isCast: false, conversionGroupOpt: null, optLeftType0, diagnostics);
                    diagnostics.Add(node, useSiteInfo);
                    var convertedRightOperand = CreateConversion(rightOperand, rightConversion, optLeftType0, diagnostics);
                    return new BoundNullCoalescingOperator(node, leftOperand, convertedRightOperand,
                        leftPlaceholder, leftConversion, BoundNullCoalescingOperatorResultKind.LeftUnwrappedType, optLeftType0);
                }
            }
```

It generates conversion that unwraps nullable left, which results in the ```Value``` access. I think we can represent it differently, as an identity conversion on top of unwrapped prior to the conversion value:
```
            if (isLeftNullable)
            {
                var rightConversion = Conversions.ClassifyImplicitConversionFromExpression(rightOperand, optLeftType0, ref useSiteInfo);
                if (rightConversion.Exists)
                {
                    var leftPlaceholder = new BoundValuePlaceholder(leftOperand.Syntax, optLeftType0).MakeCompilerGenerated();
                    diagnostics.Add(node, useSiteInfo);
                    var convertedRightOperand = CreateConversion(rightOperand, rightConversion, optLeftType0, diagnostics);
                    return new BoundNullCoalescingOperator(node, leftOperand, convertedRightOperand,
                        leftPlaceholder, leftConversion: leftPlaceholder, BoundNullCoalescingOperatorResultKind.LeftUnwrappedType, optLeftType0);
                }
            }
```

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 3), assuming CI is passing

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:19278]: Do we exercise this adequately already with the 'TestUsedAssemblies' leg?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:19329]: Do we need this check? If we do, should we verify what exactly the emit diagnostics are?

File: src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/OperatorTests.cs:4840]: Similarly: is this needed when we have the VerifyUsedAssemblies leg? And if we do need it, should we verify the set of diagnostics here? Same nit for the below calls to 'GetEmitDiagnostics'.

**Review by RikkiGibson — APPROVED** [review]
LGTM with some nits

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:19329]: That check was throwing/asserting before. Technically, that check is now embedded in `.VerifyDiagnostics()` in the new branch, but I wanted to make it explicit in the tests to fix the issue.

**Pull request merged**
